src/otx/v2/adapters/torch/mmengine/mmdet/modules/models/heads/custom_atss_head.py

```diff
@@ -0,0 +1,282 @@
+        return {"loss_cls": losses_cls, "loss_bbox": losses_bbox, "loss_centerness": loss_centerness}
+
+    def loss_by_feat_single(
+        self,
+        anchors: Tensor,
+        cls_score: Tensor,
+        bbox_pred: Tensor,
+        centerness: Tensor,
+        labels: Tensor,
+        label_weights: Tensor,
+        bbox_targets: Tensor,
+        valid_label_mask: Tensor,
+        avg_factor: float,
+    ) -> tuple[Tensor, Tensor, Tensor, Tensor]:
+        """Compute loss of a single scale level.
+
+        Args:
+            cls_score (Tensor): Box scores for each scale level
+                Has shape (N, num_anchors * num_classes, H, W).
+            bbox_pred (Tensor): Box energies / deltas for each scale
+                level with shape (N, num_anchors * 4, H, W).
+            centerness(Tensor): Centerness scores for each scale level.
+            anchors (Tensor): Box reference for each scale level with shape
+                (N, num_total_anchors, 4).
+            labels (Tensor): Labels of each anchors with shape
+                (N, num_total_anchors).
+            label_weights (Tensor): Label weights of each anchor with shape
+                (N, num_total_anchors)
+            bbox_targets (Tensor): BBox regression targets of each anchor with
+                shape (N, num_total_anchors, 4).
+            avg_factor (float): Average factor that is used to average
+                the loss. When using sampling method, avg_factor is usually
+                the sum of positive and negative priors. When using
+                `PseudoSampler`, `avg_factor` is usually equal to the number
+                of positive priors.
+            valid_label_mask (Tensor): Label mask for consideration of ignored
+                label with shape (N, num_total_anchors, 1).
+
+        Returns:
+            Tuple[Tensor, Tensor, Tensor, Tensor]: Losses.
+        """
+        anchors = anchors.reshape(-1, 4)
+        cls_score = cls_score.permute(0, 2, 3, 1).reshape(-1, self.cls_out_channels).contiguous()
+        bbox_pred = bbox_pred.permute(0, 2, 3, 1).reshape(-1, 4)
+        centerness = centerness.permute(0, 2, 3, 1).reshape(-1)
+        bbox_targets = bbox_targets.reshape(-1, 4)
+        labels = labels.reshape(-1)
+        label_weights = label_weights.reshape(-1)
+        valid_label_mask = valid_label_mask.reshape(-1, self.cls_out_channels)
+
+        # FG cat_id: [0, num_classes -1], BG cat_id: num_classes
+        pos_inds = self._get_pos_inds(labels)
+
+        if self.use_qfl:
+            quality = label_weights.new_zeros(labels.shape)
+
+        if len(pos_inds) > 0:
+            pos_bbox_targets = bbox_targets[pos_inds]
+            pos_bbox_pred = bbox_pred[pos_inds]
+            pos_anchors = anchors[pos_inds]
+            pos_centerness = centerness[pos_inds]
+
+            centerness_targets = self.centerness_target(pos_anchors, pos_bbox_targets)
+            if self.reg_decoded_bbox:
```

Comment:
same as above, not sure what `reg` means above

Reply:
I think it's regression. And unfortunately, we cannot change this name, since it comes from mmdet

Reply:
Thanks for the clarification. When I read the mmcv-based code, I feel like I'm trying to decode cryptography :)